File: leetcode/114.py
# -*- coding: utf-8 -*-
# @Author: gunjianpan
# @Date:   2020-08-02 00:25:21
# @Last Modified by:   gunjianpan
# @Last Modified time: 2020-08-02 00:26:26

"""
144. Binary Tree Preorder Traversal Medium
Given a binary tree, return the preorder traversal of its nodes' values.

Example:

Input: [1,null,2,3]
   1
    \
     2
    /
   3

Output: [1,2,3]
Follow up: Recursive solution is trivial, could you do it iteratively?

Accepted 506,529 Submissions 911,624
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """

        def dfs(head: TreeNode):
            if head is None:
                return
            lists.append(head)
            dfs(head.left)
            dfs(head.right)

        lists = []
        dfs(root)
        N = len(lists)
        for ii in range(1, N):
            prev, now = lists[ii - 1], lists[ii]
            prev.left = None
            prev.right = now

File: leetcode/114.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  public void nlr(TreeNode root, TreeNode pre) {
    if (root != null) {
      pre.right = root;
      pre.left = null;
      pre = pre.right;
      System.out.println(pre.val + " " + root.val);

      nlr(root.left, pre);
      nlr(root.right, pre);
    }
  }
  public void flatten(TreeNode root) {
    TreeNode pre = new TreeNode(0);
    nlr(root, pre);
    root = pre.right;
  }
}

class Solution {
  public void flatten(TreeNode root) {
    List<TreeNode> list = new ArrayList<TreeNode>();
    preorderTraversal(root, list);
    int size = list.size();
    for (int i = 1; i < size; i++) {
      TreeNode prev = list.get(i - 1), curr = list.get(i);
      prev.left = null;
      prev.right = curr;
    }
  }

  public void preorderTraversal(TreeNode root, List<TreeNode> list) {
    if (root != null) {
      list.add(root);
      preorderTraversal(root.left, list);
      preorderTraversal(root.right, list);
    }
  }
}
